src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -4119,25 +4121,55 @@ trait Types
    *  indirectly referenced by type `tpe1`. If there are no remaining type
    *  parameters, simply returns result type `tpe`.
    */
-  def existentialAbstraction(tparams: List[Symbol], tpe0: Type, flipVariance: Boolean = false): Type =
-    if (tparams.isEmpty) tpe0
+  def existentialAbstraction(tparams: List[Symbol], tpe0: Type, flipVariance: Boolean = false): Type = {
+
+    /* We want to narrow the list of type parameters tparams to only those which are either
+     * (a) directly contained by tpe, or
+     * (b) contained by the typeInfo of another parameter from tparams, known to be referred by tpe
+     */
+    def transitiveReferredFrom(tpe: Type): List[Symbol] = tparams match {
+      case tparam :: Nil =>
+        // This is for optimisationa: should be equivalent to general one.
+        if (tpe contains tparam) tparams else Nil
+      case _ =>
+        /* The algorithm to compute the transitive closure uses several temporary lists (mutable ListBuffer)
+         *  - pending: elements from tparams not yet known to be in the transitiveClosure
+         *  - border: we know they are in closure, but we use them for search new elements
+         *  - closed: already in closure, and we already searched for new elements.
+         *
+         * Invariant: pending, closed, and border border1 are these lists always form a partition of `tparams`.
+         * Each element in tparams goes from pending to border, and from border to closed
+         */
+        val pending   = mutable.ListBuffer.empty[Symbol]
+        var border  = mutable.ListBuffer.empty[Symbol]
+        partitionInto(tparams, tpe.contains, border, pending)
+        val closed    = mutable.ListBuffer.empty[Symbol]
+        var borderAux = mutable.ListBuffer.empty[Symbol]
+        while (border.nonEmpty) {
+          borderAux.clear
```

Comment:
You could put this at the end of the loop, since it won't be needed to clear it initially.

Reply:
Looking at it, you could do `swap.clear` after `val swap = border`

Reply:
I preferred it at the start, to make it clear that the `borderAux` is empty. Note that using a `borderAux` is for memory: it would be equivalent to just declare  `val borderAux = ListBuffer.empty`, which would create a new object every iteration.